src/Libraries/LibPackedStream.sol

```diff
@@ -0,0 +1,147 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+pragma solidity ^0.8.17;
+
+/// @title LibPackedStream
+/// @author LI.FI (https://li.fi)
+/// @notice Minimal byte-stream reader for compact calldata formats
+/// @custom:version 1.0.0
+library LibPackedStream {
+    /// @dev Returns the start and finish pointers for a bytes array.
+    function _bounds(
+        bytes memory data
+    ) private pure returns (uint256 start, uint256 finish) {
+        assembly {
+            start := add(data, 32)
+            finish := add(start, mload(data))
+        }
+    }
+
+    /** @notice Creates stream from data
+     * @param data data
+     */
+    function createStream(
+        bytes memory data
+    ) internal pure returns (uint256 stream) {
+        (uint256 start, uint256 finish) = _bounds(data);
+        assembly {
+            stream := mload(0x40)
+            mstore(stream, start)
```

Comment:
Do we not need to also use diamond storage here to prevent memory clashes?

Reply:
what do you mean by memory clashes? it won't affect storage for sure as it uses `mstore` not `sstore` so in this we are safe. Do you see any other memory clashes?